Source/Particles/Deposition/CurrentDeposition.H

```diff
@@ -1533,6 +1533,670 @@ void doChargeConservingDepositionShapeNImplicit (const amrex::ParticleReal * con
+                        }
+                    }
+                }
+
+                // deposit Jy for this segment
+                amrex::Real this_Jy;
+                for (int i=0; i<=depos_order; i++) {
+                    for (int j=0; j<=depos_order-1; j++) {
+                        for (int k=0; k<=depos_order; k++) {
+                            this_Jy = wqy*sy_cell[j]*( sx_old_node[i]*sz_old_node[k]*one_third
+                                                     + sx_old_node[i]*sz_new_node[k]*one_sixth
+                                                     + sx_new_node[i]*sz_old_node[k]*one_sixth
+                                                     + sx_new_node[i]*sz_new_node[k]*one_third )*seg_factor_y;
+                            amrex::Gpu::Atomic::AddNoRet( &Jy_arr(lo.x+i0_node+i, lo.y+j0_cell+j, lo.z+k0_node+k), this_Jy);
+                        }
+                    }
+                }
+
+                // deposit Jz for this segment
+                amrex::Real this_Jz;
+                for (int i=0; i<=depos_order; i++) {
+                    for (int j=0; j<=depos_order; j++) {
+                        for (int k=0; k<=depos_order-1; k++) {
+                            this_Jz = wqz*sz_cell[k]*( sx_old_node[i]*sy_old_node[j]*one_third
+                                                     + sx_old_node[i]*sy_new_node[j]*one_sixth
+                                                     + sx_new_node[i]*sy_old_node[j]*one_sixth
+                                                     + sx_new_node[i]*sy_new_node[j]*one_third )*seg_factor_z;
+                            amrex::Gpu::Atomic::AddNoRet( &Jz_arr(lo.x+i0_node+i, lo.y+j0_node+j, lo.z+k0_cell+k), this_Jz);
+                        }
+                    }
+                }
+
+                // update old segment values
+                if (ns < num_segments-1) {
+                    x0_old = x0_new;
+                    y0_old = y0_new;
+                    z0_old = z0_new;
+                }
+
+            } // end loop over segments
+
+#elif defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+
+            // compute cell crossings in X-direction
+            const auto i_old = static_cast<int>(x_old-shift);
+            const auto i_new = static_cast<int>(x_new-shift);
+            int cell_crossings_x = std::abs(i_new-i_old);
+            num_segments += cell_crossings_x;
+
+            // compute cell crossings in Z-direction
+            const auto k_old = static_cast<int>(z_old-shift);
+            const auto k_new = static_cast<int>(z_new-shift);
+            int cell_crossings_z = std::abs(k_new-k_old);
+            num_segments += cell_crossings_z;
+
+            // need to assert that the number of cell crossings in each direction
+            // is within the range permitted by the number of guard cells
+            // e.g., if (num_segments > 3) ...
```

Comment:
Debug-mode only assert would be:
```suggestion
            // e.g., if (num_segments > 3) ...
            AMREX_ASSERT_WITH_MESSAGE(
                num_segments < 3,
                "VB Deposition: Detected particle with too many Cell Crossings!"
            );
```

Reply:
Will keep this in mind. It really should be based on the number of guard cells rather than a hardwired value. So I think I will defer this change to the future when the implicit method is more mature.

Reply:
Sounds good! We have two kinds of asserts: the one here is only added in debug mode - the `WARPX_ALWAYS_ASSERT` is always run and should not be used in hot loops like GPU kernels.

Reply:
Good to know. Thanks.